File: c/questions/maxPathSum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#define MAX(a, b) ((a) > (b) ? (a) : (b))

int dfsS(struct TreeNode* node, int* gMax) {
    if (node == NULL) {
        return 0;
    }
    int l = dfsS(node->left, gMax);
    l = MAX(l ,0);
    int r = dfsS(node->right, gMax);
    r = MAX(r, 0);
    int cgm = node->val + l + r;
    *gMax = MAX(*gMax, cgm);
    int ctm = node->val + MAX(l, r);
    return ctm;
}

int maxPathSum(struct TreeNode* root) {
    int gm = INT_MIN;
    dfsS(root, &gm);
    return gm;
}

File: javascript/questions/maxPathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    let maxSum = Number.NEGATIVE_INFINITY;
    const dfs = function(node) {
	if (node === null) {
	    return 0;
	}
	const l = Math.max(dfs(node.left), 0);
	const r = Math.max(dfs(node.right), 0);
	maxSum = Math.max(maxSum, node.val+l+r);
	return Math.max(l, r) + node.val;
    }
    dfs(root);
    return maxSum;
};
